### PR TITLE
Render the audio

### PR DESCRIPTION
**Pull request by padenot**
This works well for me in Chrome 94 beta with Web Codecs enabled on my Linux box, here's how it goes:

- A `SharedArrayBuffer`-based ring buffer is given to a very simple `AudioWorkletProcessor` (that reads from it) and the control thread (for now the main thread, but we'll be able to move that to a Web Worker)
- The `AudioRenderer` pre-buffers some audio and writes the PCM to this ring buffer, on initialization.
- When the `AudioContext` starts, the `AudioWorkletProcessor` starts to read from this ring buffer to output the audio.
- The `AudioRenderer` sees that the ring buffer becomes empty and crosses its threshold for decoding -- it then proceeds to decode some compressed audio and pushes the PCM to the ring buffer. If the threshold isn't reached, it simply re-schedules itself in the future.
- Pausing the `AudioRenderer` is made via `suspend`ing the `AudioContext`, so pause/play is supported
- A volume slider has been added: this is just a `GainNode` and an `<input type=range>` -- we get smooth volume ramps for free :-).
- The clock is simply the `AudioContext.currentTime`, offset by the output latency, converted to microseconds.

But it is only a first cut, this is what I'm planning to fix:
- [x] Only the left channel is rendered for now, because that was easier enough to get something going. Not sure if Chrome has all the `copyTo` variants already, and `[AllowShared]`, but in any case I have code lying around to interleave/deinterleave if necessary
- [x] There is a copy we should skip in the decoder output callback
- [ ] A/V sync is not as good as it could be because of the lack of `AudioContext.outputLatency` in Chrome, see https://blog.paul.cx/post/audio-video-synchronization-with-the-web-audio-api/ for details. For now, I picked a number that was kind of OK for my Linux workstation with my sound card, but YMMV. We'll do something fancier, per platform, in a second step, so it looks good
- [x] Increase the latency of the `AudioContext`: no need for extreme low latency here.

All the code is in PR is from me originally so no license issue. The ring buffer and the server are from https://github.com/padenot/ringbuf.js, but I've just copied/tweaked the necessary files for this demo.

### CODE REVIEW

**Comment by padenot** [comment]
> Only the left channel is rendered for now, because that was easier enough to get something going. Not sure if Chrome has all the `copyTo` variants already, and `[AllowShared]`, but in any case I have code lying around to interleave/deinterleave if necessary

I was using the wrong `google-chrome-unstable` it seems like, I took another one (95) and it doesn't throw anymore.

File: audiosink.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+registerProcessor("AudioSink", class AudioSink extends AudioWorkletProcessor {
+  constructor(options) {
+    super();
+    let sab = options.processorOptions.sab;
+    this.consumerSide = new RingBuffer(sab, Float32Array);
+  }
+  process(inputs, outputs, params) {
+    // Assuming mono for now
+    var available_read = this.consumerSide.available_read();
```
Comment (chcunningham) [inline audiosink.js:9]: This seems unused.

**Review by chcunningham — APPROVED** [review]
Looks great. I had some questions, but feel free to submit whenever you're happy.

File: audio_renderer.js [hunk 2]
```diff
@@ -23,65 +37,110 @@ export class AudioRenderer {
...
+    return this.audioContext.suspend();
   }
 
   getMediaTime() {
-    return this.lastRenderedMediaTimestamp;
+    let totalOutputLatency = 0.0;
+    if (this.audioContext.outputLatency == undefined) {
+      // Put appropriate values for Chromium here, not sure what latencies are
+      // used. Likely OS-dependent, certainly hardware dependant. Assume 40ms.
+      totalOutputLatency += 0.04;
```
Comment (chcunningham) [inline audio_renderer.js:120]: @hoch FYI. bumping https://bugs.chromium.org/p/chromium/issues/detail?id=785409

**Comment by guest271314** [comment]
If I understand the requirement correctly this https://guest271314.github.io/webcodecs/ is how I serialize and deserialize audio with WebCodecs. It is non-trivial resampling the fixed 48000 sample rate that Chromium has hardcoded, especially when the input file is 22050, 1 channel, and I never set Opus encoder sample rate to 48000.

File: audio_renderer.js [hunk 3]
```diff
@@ -90,17 +149,26 @@ export class AudioRenderer {
...
+  availableWrite() {
+    return this.ringbuffer.available_write() / this.sampleRate;
+  }
 
+  async fillDataBuffer() {
+    let inBuffer = this.availableWrite();
+    if (inBuffer < DATA_BUFFER_TARGET_DURATION ||
+        this.decoder.decodeQueueSize > DECODER_QUEUE_SIZE_MAX) {
+      // Check back later
+      window.setTimeout(this.fillDataBuffer.bind(this), 10);
```
[the remark this answers is not in the document]
Comment (guest271314) [reply audio_renderer.js:162]: The last function wanted in any audio processing code is `setTimeout()`. That means inaccuracy is afoot.

**Comment by guest271314** [comment]
> The AudioRenderer sees that the ring buffer becomes empty and crosses its threshold for decoding -- it then proceeds to decode some compressed audio and pushes the PCM to the ring buffer. If the threshold isn't reached, it simply re-schedules itself in the future.

The rescheduling part can be problematic re glitches and gaps between schedules. One option is to utilize `WebAssembly.Memory.grow()` to dynamically grow the underlying buffer - within the 4GB Chromium limitation - to write and read to a single contigous block of memory, something like

    const initial = (384 * 512 * 3) / 65536; // 3 seconds
    const maximum = (384 * 512 * 60 * 60) / 65536; // 1 hour
    let started = false;
    let readOffset = 0;
    let init = false;
    const memory = new WebAssembly.Memory({
      initial,
      maximum,
      shared: true,
    });
    console.log(memory.buffer.byteLength, initial / 65536);
    // ...
    await readable.pipeTo(
      new WritableStream(
        {
          start() {
            console.log('writable start');
          },
          async write(value, controller) {
            console.log(value, value.byteLength, memory.buffer.byteLength);
            if (readOffset + value.byteLength > memory.buffer.byteLength) {
              console.log('before grow', memory.buffer.byteLength);
              memory.grow(3);
              console.log('after grow', memory.buffer.byteLength);
            }
            let uint8_sab = new Uint8Array(memory.buffer);
            let i = 0;
            if (!init) {
              init = true;
              i = 44;
            }
            for (; i < value.buffer.byteLength; i++, readOffset++) {
              if (readOffset + 1 >= memory.buffer.byteLength) {
                console.log(`memory.buffer.byteLength before grow() for loop: ${memory.buffer.byteLength}.`);
                memory.grow(3);
                console.log(`memory.buffer.byteLength after grow() for loop: ${memory.buffer.byteLength}`);
                uint8_sab = new Uint8Array(memory.buffer);
              }              
              uint8_sab[readOffset] = value[i];
            }
            if (!started) {
              started = true;
              aw.port.postMessage({
                started: true,
              });
            }
          },
          close() {
            console.log('writable', readOffset, memory.buffer.byteLength);
            aw.port.postMessage({
              readOffset,
            });
          },
        }
      )
    );

File: ringbuf.js [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+//
+// The producer and the consumer can be on separate threads, but cannot change roles,
+// except with external synchronization.
+
+export class RingBuffer {
+  static getStorageForCapacity(capacity, type) {
+    if (!type.BYTES_PER_ELEMENT) {
+      throw "Pass in a ArrayBuffer subclass";
+    }
+    var bytes = 8 + (capacity + 1) * type.BYTES_PER_ELEMENT;
```
[the remark this answers is not in the document]
Comment (padenot) [reply ringbuf.js:11]: This is explained line 26. It's necessary to distinguish an empty and full ring buffer.

File: server.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+      response.writeHead(200, { "Content-Type": mimeType ,
+        "Cross-Origin-Opener-Policy": "same-origin",
+        "Cross-Origin-Embedder-Policy": "require-corp" });
+      response.write(file, "binary");
+      response.end();
+    });
+  });
+}).listen(parseInt(port, 10));
+
+console.log("Static file server running at\n  => https://localhost:" + port + "/\nCTRL + C to shutdown");
```
[the remark this answers is not in the document]
Comment (padenot) [reply server.js:57]: Yes, this used to be https, I missed this.

File: audiosink.js (hunk 1 above)
Comment (padenot) [reply audiosink.js:9]: Yes, debugging leftovers.

File: audio_renderer.js [hunk 6]
```diff
@@ -23,65 +37,110 @@ export class AudioRenderer {
...
+    console.assert(AudioDecoder.isConfigSupported(config));
     this.decoder.configure(config);
 
+    // Initialize the AudioWorkletProcessor
+    this.audioContext = new AudioContext({ sampleRate: trackInfo.sampleRate });
+    this.audioContext.suspend();
+    // Initialize the ring buffer between the decoder and the real-time audio
+    // rendering thread. The AudioRenderer has buffer space for approximately
+    // 200ms of decoded audio ahead, but targets 100ms.
+    let sampleCountIn500ms =
```
[the remark this answers is not in the document]
Comment (padenot) [reply audio_renderer.js:62]: Arbitrary for now, I haven't done any tuning, and 500ms is enough.

File: audio_renderer.js [hunk 7]
```diff
@@ -23,65 +37,110 @@ export class AudioRenderer {
...
+
+    console.assert(AudioDecoder.isConfigSupported(config));
     this.decoder.configure(config);
 
+    // Initialize the AudioWorkletProcessor
+    this.audioContext = new AudioContext({ sampleRate: trackInfo.sampleRate });
+    this.audioContext.suspend();
+    // Initialize the ring buffer between the decoder and the real-time audio
+    // rendering thread. The AudioRenderer has buffer space for approximately
+    // 200ms of decoded audio ahead, but targets 100ms.
```
[the remark this answers is not in the document]
Comment (padenot) [reply audio_renderer.js:61]: Yes, I found a few chunks uncommited in my repo this morning that should have gone in, I'll fix.

File: audio_renderer.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (padenot) [reply audio_renderer.js:162]: When using an SPSC buffer, it's enough to look at the fullness of the buffer. I suppose we could use `atomics.wait()/notify()`, but it's unnecessary. This was a _very_ crude buffering strategy, what's coming is more appropriate.

This isn't audio processing code, it's audio decoding, btw.

**Comment by padenot** [comment]
What's left to do is to check A/V sync on various OSes, I have only been testing on a Linux desktop with a USB DAC.

**Pull request merged**